### ISSUE TITLE
three questions when using pksensi

### ISSUE THREAD

**Issue by Yoqing** [nanhung/pksensi#10]
Dear Nan-Hung,

Thanks for the amazing package. 
I met an error when using it. But I have no idea how to solve it. Could you please give me some suggestions on it? Thanks in advance~~~

**the error is 
Starting time: 2023-11-07 11:25:45.597959
DLSODE-  At T (=R1) and step size H (=R2), the    
      corrector convergence failed repeatedly     
      or with ABS(H) = HMIN   
In above message, R1 = 0, R2 = 3.8147e-08
 
Error in tmp[k + 1, vars[l]] : subscript out of bounds
In addition: Warning messages:
1: In lsode(y, times, func, parms, ...) :
  repeated convergence test failures on a step, but integration was successful - inaccurate Jacobian matrix?
2: In lsode(y, times, func, parms, ...) :
  Returning early. Results are accurate, as far as they go**

@My code is here, if it needed:
##the tv of parameters
Qecf = 0.2/1000;
Qcsf = 2.2/1000;
Vbr = 1.73;
Vlv = 50/1000;
Vtfv = 50/1000;
Vcm = 17/1000;
Vsas = 180/1000;

CL = 15.41; 
Vcen = 60.34;
Q = 24.27;
CLin = 212.92/1000;
CLout = 211.7/1000;
CLpl = 1.89/1000;
CLlp = 3.48/1000;
CLpt = 29.01/1000;
CLtp = 68.12/1000;
Vper = 151.86;

# omega 
omega_para <- data.frame(
  omega_Vcen = 16.9,
  omega_CL = 4.58,
  omega_Vper = 4.14,
  omega_Q = 15,
  omega_CLin = 23,
  omega_CLout = 5.67,
  omega_CLlp = 15.6,
  omega_CLpl = 17.9,
  omega_CLtp = 23.4,
  omega_CLpt = 7.76
)

# calculate sd 
{
sd_Vcen = sqrt(exp(omega_para$omega_Vcen^2)-1)
sd_CL = sqrt(exp(omega_para$omega_CL^2)-1)
sd_Vper = sqrt(exp(omega_para$omega_Vper^2)-1)
sd_Q = sqrt(exp(omega_para$omega_Q^2)-1)
sd_CLin = sqrt(exp(omega_para$omega_CLin^2)-1)
sd_CLout = sqrt(exp(omega_para$omega_CLout^2)-1)
sd_CLlp = sqrt(exp(omega_para$omega_CLlp^2)-1)
sd_CLpl = sqrt(exp(omega_para$omega_CLpl^2)-1)
sd_CLtp = sqrt(exp(omega_para$omega_CLtp^2)-1)
sd_CLpt = sqrt(exp(omega_para$omega_CLpt^2)-1)
}

# initial value 
yini <- unlist(c(data.frame(
  Acen = 15*0.25*1^6,       # Quantity in plasma compartment (ng)
  Aper = 0.0,       # ~           peripheral compartment
  Aecf = 0.0,       # ~           ECF
  Alv = 0.0,        # ~         
  Atfv = 0.0,       # ~         
  Acm = 0.0,        # ~       
  Asas = 0.0 
)))

PBPKmodAPAP <- function(t,state,parameters){
  with(as.list(c(state,parameters)), {
    Ccen <-  Acen/Vcen
    Cper <- Aper/Vper
    Cecf <-  Aecf/Vbr
    Clv <-  Alv/Vlv
    Ctfv <-  Atfv/Vtfv
    Ccm <-  Acm/Vcm
    Csas <-  Asas/Vsas
    
    
    #ODE equations
    dAcen = Q * Cper - Q * Ccen - CL * Ccen - CLin * Ccen + CLout * Cecf + Clv * CLlp - Ccen * CLpl + Ctfv * CLtp - Ccen * CLpt + Qcsf * Csas + 0.02/100*200000
    dAper = Q * Ccen - Q * Cper
    dAecf = CLin * Ccen -CLout * Cecf - Qecf * Cecf
    dAlv = Qecf * Cecf - Clv * Qcsf - CLlp * Clv + CLpl * Ccen
    dAtfv = Qcsf* Clv - Ctfv * Qcsf + CLpt *Ccen - CLtp * Ctfv
    dAcm = Qcsf* Ctfv - Ccm * Qcsf
    dAsas = Qcsf* Ccm - Csas * Qcsf;
    
        list(c(dAcen, dAper, dAecf, dAlv, dAtfv, dAcm, dAsas), 
             "Ccen" = Ccen, "Cecf" = Cecf, "Clv" = Clv, "Ccm" = Ccm) 
      })
    }

outputs = c("Ccen","Cecf","Clv","Ccm")  
results <- ode(y=yini, times=times, func=PBPKmodAPAP, parms=para, method="lsoda")
APAPamount <- as.data.frame(results)
q <- c("qunif", "qunif","qunif","qunif","qunif","qunif","qunif", "qlnorm","qlnorm","qlnorm","qlnorm","qlnorm","qlnorm","qlnorm","qlnorm","qlnorm")
params <- c("Qecf", "Qcsf", "Vbr","Vlv", "Vtfv", "Vcm", "Vsas", "Vcen", "CL","Q","CLin","CLout","CLpl","CLlp","CLpt","CLtp")
LL <- 0.1 # 10% lower limit
UL <- 1.1.
q.arg <- list(
  list(min = para["Qecf"]*LL, max= para["Qecf"]*UL),       ##uniform with bounds corresponding to LL and UL of the nominal value
  list(min = para["Qcsf"]*LL, max= para["Qcsf"]*UL),
  list(min = para["Vbr"]*LL, max =  para["Vbr"]*UL),
  list(min = para["Vlv"]*LL, max =  para["Vlv"]*UL),
  list(min = para["Vtfv"]*LL, max =  para["Vtfv"]*UL),
  list(min = para["Vcm"]*LL, max = para["Vcm"]*UL),
  list(min = para["Vsas"]*LL, max = para["Vsas"]*UL),
  list(meanlog = log(max(para["Vcen"], 1e-10)), sdlog = log(sd_Vcen)),        ##lognormal distribution with CV from Monolix
  list(meanlog = log(max(para["CL"], 1e-10)), sdlog = log(sd_CL)),
  list(meanlog = log(max(para["Q"], 1e-10)), sdlog = log(sd_Q)),
  list(meanlog = log(max(para["CLin"], 1e-10)), sdlog = log(sd_CLin)),
  list(meanlog = log(max(para["CLout"], 1e-10)), sdlog = log(sd_CLout)),
  list(meanlog = log(max(para["CLpl"], 1e-10)), sdlog = log(sd_CLpl)),
  list(meanlog = log(max(para["CLlp"], 1e-10)), sdlog = log(sd_CLlp)),
  list(meanlog = log(max(para["CLpt"], 1e-10)), sdlog = log(sd_CLpt)),
  list(meanlog = log(max(para["CLtp"], 1e-10)), sdlog = log(sd_CLtp)))
set.seed(1234)
length(params)==length(q)
x <- rfast99(params = params, n = 20, q = q, q.arg = q.arg, rep = 1)

dim(x$a)
out <- solve_fun(x, time=times, func = PBPKmodAPAP, initState = yini, outnames = outputs)  # error occurred here

Thanks again
Best wishes,
Yuqing

**Comment by nanhung** [comment]
Yuqing,

Two issues I found in your code are,
- The sample size of efast was set to 20 which is not enough to generate the sample number for sensitivity analysis.
- The sd defined in the original code is not appropriate in the parameter setting of lognormal distribution. This is why you got the error message.

I wrote the example for you. Feel free to modify the code based on your requirements.
https://rpubs.com/Nanhung/pksensi110723
